### PR TITLE
Update model with new datasets, pathologies and contrasts

### LINKED ISSUE #121: open PR to merge code to generate nice figure
https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/aff61ad8e014035e9174c69fce16cc7fe119f5e1/monai/utils.py#L66

### PR DESCRIPTION
**Pull request by naga-karthik**
This PR brings the changes to the codebase used for training a newer version of the model with more datasets, contrasts and pathologies. Info about all datasets used, pathologies included, pathology-contrast-wise splits, resolutions, etc. can be found in the drop-down below. 

<details>

<summary> (mega) dataset stats </summary>

DATASETS USED FOR MODEL TRAINING (n=15):
	- basel-mp2rage
	- canproco
	- data-multi-subject
	- dcm-brno
	- dcm-zurich-lesions-20231115
	- dcm-zurich-lesions
	- dcm-zurich
	- lumbar-epfl
	- lumbar-vanderbilt
	- nih-ms-mp2rage
	- sci-colorado
	- sci-paris
	- sci-zurich
	- sct-testing-large
	- spider-challenge-2023

SUBJECT-WISE PATHOLOGY SPLIT:

| Pathology   |   Number of Subjects |
|:------------|---------------------:|
| ALS         |                   32 |
| DCM         |                  384 |
| HC          |                  428 |
| LBP         |                  218 |
| MS          |                  524 |
| NMO         |                   19 |
| PPMS        |                   65 |
| RIS         |                   61 |
| RRMS        |                  318 |
| SCI         |                  196 |
| SPMS        |                   14 |
| SYR         |                    2 |
| TOTAL       |                 2261 |

CONTRAST-WISE PATHOLOGY SPLIT (a subject can have multiple contrasts):

|         |   HC |   MS |   RIS |   RRMS |   PPMS |   DCM |   SPMS |   SCI |   NMO |   SYR |   ALS |   LBP |   #total_per_contrast |
|:--------|-----:|-----:|------:|-------:|-------:|------:|-------:|------:|------:|------:|------:|------:|----------------------:|
| dwi     |  184 |    1 |     0 |      0 |      0 |    69 |      0 |     4 |     0 |     1 |     5 |     0 |                   264 |
| mp2rage |    0 |    0 |     0 |    138 |     10 |     0 |     28 |     0 |     0 |     0 |     0 |     0 |                   176 |
| mt-off  |  184 |    0 |     0 |      0 |      0 |    59 |      0 |     0 |     0 |     0 |     0 |     0 |                   243 |
| mt-on   |  184 |    0 |     0 |      0 |      0 |    65 |      0 |     0 |     0 |     0 |     5 |     0 |                   254 |
| psir    |   42 |    0 |    54 |    193 |     44 |     0 |      0 |     0 |     0 |     0 |     0 |     0 |                   333 |
| stir    |   10 |    0 |     7 |     56 |     16 |     0 |      0 |     0 |     0 |     0 |     0 |     0 |                    89 |
| t1map   |    0 |    0 |     0 |     69 |      5 |     0 |     14 |     0 |     0 |     0 |     0 |     0 |                    88 |
| t1w     |  249 |   27 |     0 |      0 |      0 |    61 |      0 |     0 |    19 |     1 |     0 |   196 |                   553 |
| t2star  |  237 |  452 |     0 |      0 |      0 |   146 |      0 |     4 |     0 |     1 |    27 |     0 |                   867 |
| t2w     |  252 |  552 |    61 |    229 |     57 |   476 |      0 |   261 |     0 |     1 |    32 |   251 |                  2172 |
| unit1   |   53 |   50 |     0 |     69 |      5 |     0 |     14 |     0 |     0 |     0 |     0 |     0 |                   191 |
| TOTAL   | 1395 | 1082 |   122 |    754 |    137 |   876 |     56 |   269 |    19 |     4 |    69 |   447 |                  5230 |

SPLITS ACROSS DIFFERENT CONTRASTS (n=11):

| contrast   |   train |   validation |   test |   #images_per_contrast |
|:-----------|--------:|-------------:|-------:|-----------------------:|
| dwi        |     175 |           38 |     51 |                    264 |
| mp2rage    |     142 |           16 |     18 |                    176 |
| mt-off     |     158 |           36 |     49 |                    243 |
| mt-on      |     168 |           36 |     50 |                    254 |
| psir       |     217 |           49 |     67 |                    333 |
| stir       |      56 |           15 |     18 |                     89 |
| t1map      |      71 |            8 |      9 |                     88 |
| t1w        |     411 |           64 |     78 |                    553 |
| t2star     |     638 |          113 |    116 |                    867 |
| t2w        |    1623 |          261 |    288 |                   2172 |
| unit1      |     152 |           19 |     20 |                    191 |
| TOTAL      |    3811 |          655 |    764 |                   5230 |

RESOLUTIONS PER CONTRAST PER ORIENTATION (in mm^3):

How to interpret the table: Each row corresponds to the contrast and its orientation with the median, min, max and mean resolutions in mm^3.
For simplification, if a contrast does not have any images in a particular orientation in the dataset, then the row is not present in the table.
For e.g. if you want to report the mean (min, max) resolution of a contrast, say, 'dwi_axial', 
then you pick the respective element in each of the columns.
	 i.e. mean in-plane resolution: 0.89x0.89; range: (0.34x0.34, 1x1), and, likewise, Slice thickness: 5.1; range: (4, 17.5)

| contrast          | median_resolution_rpi   | min_resolution_rpi   | max_resolution_rpi   | mean_resolution_rpi   |
|:------------------|:------------------------|:---------------------|:---------------------|:----------------------|
| mp2rage_sagittal  | 1.2 x 1.0 x 1.0         | 1.2 x 1.0 x 1.0      | 1.2 x 1.0 x 1.0      | 1.2 x 1.0 x 1.0       |
| psir_sagittal     | 3.0 x 0.69 x 0.69       | 3.0 x 0.67 x 0.67    | 3.0 x 0.69 x 0.69    | 3.0 x 0.69 x 0.69     |
| stir_sagittal     | 3.0 x 0.7 x 0.7         | 3.0 x 0.7 x 0.7      | 3.0 x 0.7 x 0.7      | 3.0 x 0.7 x 0.7       |
| t1map_sagittal    | 1.2 x 1.0 x 1.0         | 1.2 x 1.0 x 1.0      | 1.2 x 1.0 x 1.0      | 1.2 x 1.0 x 1.0       |
| t1w_sagittal      | 3.31 x 0.62 x 0.63      | 1.0 x 0.25 x 0.08    | 4.86 x 1.06 x 1.23   | 3.68 x 0.66 x 0.64    |
| t2w_sagittal      | 2.75 x 0.57 x 0.52      | 0.8 x 0.25 x 0.08    | 9.63 x 1.0 x 1.1     | 2.4 x 0.57 x 0.57     |
| unit1_sagittal    | 1.2 x 1.0 x 1.0         | 1.2 x 1.0 x 1.0      | 1.2 x 1.0 x 1.0      | 1.2 x 1.0 x 1.0       |
| dwi_axial         | 0.9 x 0.9 x 5.0         | 0.34 x 0.34 x 4.0    | 1.0 x 1.0 x 17.5     | 0.89 x 0.89 x 5.13    |
| mt-off_axial      | 0.9 x 0.9 x 5.0         | 0.68 x 0.68 x 5.0    | 0.9 x 0.9 x 5.0      | 0.89 x 0.89 x 5.0     |
| mt-on_axial       | 0.9 x 0.9 x 5.0         | 0.6 x 0.6 x 5.0      | 0.9 x 0.9 x 9.3      | 0.88 x 0.88 x 5.07    |
| t1w_axial         | 0.35 x 0.35 x 2.5       | 0.35 x 0.35 x 2.5    | 0.8 x 0.8 x 5.0      | 0.41 x 0.41 x 2.87    |
| t2star_axial      | 0.47 x 0.47 x 5.0       | 0.29 x 0.29 x 2.5    | 0.8 x 0.8 x 9.8      | 0.43 x 0.43 x 4.41    |
| t2w_axial         | 0.5 x 0.5 x 3.6         | 0.29 x 0.29 x 1.0    | 0.8 x 1.0 x 7.5      | 0.51 x 0.52 x 3.67    |
| mp2rage_isotropic | 1.0 x 1.0 x 1.0         | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0       |
| t1map_isotropic   | 1.0 x 1.0 x 1.0         | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0       |
| t1w_isotropic     | 1.0 x 1.0 x 1.0         | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0       |
| t2w_isotropic     | 0.8 x 0.8 x 0.8         | 0.8 x 0.8 x 0.8      | 1.0 x 1.0 x 1.0      | 0.81 x 0.81 x 0.81    |
| unit1_isotropic   | 1.0 x 1.0 x 1.0         | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0      | 1.0 x 1.0 x 1.0       |

MODEL NAME (for dev purposes):

	nnunet-plain_seed=50_ndata=15_ncont=11_nf=384_opt=adam_lr=0.001_AdapW_bs=2_20241018-0508

</details>

Resolves #121 #102

### CODE REVIEW

File: scripts/generate_tables.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    "t1w": ["T1w", "space-other_T1w", "acq-lowresSag_T1w"],
+    "t2w": ["T2w", "space-other_T2w", "acq-lowresSag_T2w", "acq-highresSag_T2w"],
+    "t2star": ["T2star", "space-other_T2star"],
+    "dwi": ["rec-average_dwi", "acq-dwiMean_dwi"],
+    "mt-on": ["flip-1_mt-on_space-other_MTS", "acq-MTon_MTR"],
+    "mt-off": ["flip-2_mt-off_space-other_MTS"],
+    "unit1": ["UNIT1"],
+    "psir": ["PSIR"],
+    "stir": ["STIR"]
+}
```
Comment (valosekj) [inline scripts/generate_tables.py:18]: Very minor comment.
This dict looks similar/identical to: https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/f9e58332434a11ebb798bcdc4d0e99dfa1663933/monai/main.py#L29-L41
For future scripts and PRs, you could maybe define such common dicts in a utils script and import them.

File: datasplits/datasplit_dcm-brno_seed50.yaml [hunk 2]
```diff
@@ -0,0 +1,156 @@
+test:
+- sub-1860B6472B
+- sub-2295B4676B
+- sub-2336B
+- sub-2347B6454B
```
Comment (valosekj) [inline datasplits/datasplit_dcm-brno_seed50.yaml:5]: Just to confirm, @naga-karthik. `dcm-brno` contains both HC and MildCompreession/DCM. 
Subjects with a single session (i.e., short participant_id), e.g. `sub-2336B` are HC. While subjects with two sessions (i.e., long participant_id), e.g., `sub-2347B6454B` are MildCompreession/DCM.
Are you taking this into account when creating your "(mega) dataset stats" in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/125#issue-2655187987, aren't you?

**Review by valosekj — APPROVED** [review]
Looks good, @naga-karthik! Thank you!

(note that given the high number of changes, I did only a high-level review)

File: datasplits/datasplit_dcm-brno_seed50.yaml (hunk 2 above)
Comment (naga-karthik) [reply datasplits/datasplit_dcm-brno_seed50.yaml:5]: hey @valosekj, if the corresponding `participants.tsv` has `HC` for `sub-236B` (and similarly for other short participant ids), then, yes, they are automatically taken care of!

Comment (valosekj) [reply datasplits/datasplit_dcm-brno_seed50.yaml:5]: Nice, thanks for the confirmation!

**Pull request merged**